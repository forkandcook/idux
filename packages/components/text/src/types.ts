/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export type TextStyle = 'secondary' | 'success' | 'warning' | 'danger'

export const textProps = {
  type: String as PropType<TextStyle>,
  disabled: Boolean,
  delete: Boolean,
  underline: Boolean,
  code: Boolean,
  mark: Boolean,
  keyboard: Boolean,
  strong: Boolean,
  italic: Boolean,
  copyable: Boolean,

  ellipsis: {
    type: [Boolean, Object] as PropType<boolean | EllipsisProps>,
    default: false,
  },

  // events
  onClick: Function as PropType<(e: Event) => void>,
} as const

export type TextProps = ExtractInnerPropTypes<typeof textProps>
export type TextPublicProps = ExtractPublicPropTypes<typeof textProps>
export type TextComponent = DefineComponent<Omit<HTMLAttributes, keyof TextPublicProps> & TextPublicProps>
export type TextInstance = InstanceType<DefineComponent<TextProps>>

export type EllipsisProps = Partial<{
  width: number
  rows: number
  expanded: boolean
  tooltip: string
  fontSize: number
  onExpand: () => void
}>
