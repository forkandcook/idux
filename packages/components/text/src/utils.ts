/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

let count = 0
const uuid = () => `idux-${count++}`
const contentMap = new Map<string, HTMLElement>()
const ellipsisMap = new Map<string, HTMLElement>()

function getTextWidth(text: string, style: CSSStyleDeclaration) {
  const canvas = document.createElement('canvas')
  const context = canvas.getContext('2d')!
  context.font = `${style.fontSize} ${style.fontFamily}`
  const metrics = context.measureText(text)

  return metrics.width
}

export function truncateText(container: HTMLElement, width: number, expanded = false): void {
  const id = container.dataset.id || uuid()
  if (!contentMap.has(id)) {
    contentMap.set(id, container.querySelector('span:first-child')!.cloneNode(true) as HTMLElement)
    container.dataset.id = id
  }

  if (expanded) {
    container.querySelector('span:first-child')!.outerHTML = contentMap.get(id)!.outerHTML
    return
  }

  if (ellipsisMap.has(id)) {
    container.querySelector('span:first-child')!.outerHTML = ellipsisMap.get(id)!.outerHTML
    return
  }

  const copyTag = container.querySelector('span:last-child')
  const tagStyle = getComputedStyle(container)
  const ellipsisWidth = getTextWidth('...', tagStyle)
  const copyTagWidth = copyTag ? copyTag.getBoundingClientRect().width : 0

  let remainingWidth = width - ellipsisWidth - copyTagWidth
  const nodes = Array.from(container.querySelector('span:first-child')!.childNodes)

  nodes.some(node => {
    const el = node as HTMLElement
    const text = el.textContent
    const textWidth = getTextWidth(text!, tagStyle)

    if (remainingWidth >= textWidth) {
      remainingWidth -= textWidth
      return false
    } else {
      const rate = remainingWidth / textWidth
      const content = text?.slice(0, Math.floor(text.length * rate))
      el.textContent = content + '...'
      return true
    }
  })

  if (!ellipsisMap.has(id)) {
    ellipsisMap.set(id, container.querySelector('span:first-child')!.cloneNode(true) as HTMLElement)
  }
}
