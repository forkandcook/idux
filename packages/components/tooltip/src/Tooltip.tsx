/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { TooltipProps } from './types'
import type { CSSProperties, Slots } from 'vue'

import { computed, defineComponent } from 'vue'

import { ɵOverlay } from '@idux/components/_private/overlay'
import { useGlobalConfig } from '@idux/components/config'

import { tooltipProps } from './types'
import { useTooltipOverlay } from './useTooltipOverlay'

export default defineComponent({
  name: 'IxTooltip',
  props: tooltipProps,
  setup(props, { slots, expose }) {
    const common = useGlobalConfig('common')
    const config = useGlobalConfig('tooltip')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-tooltip`)
    const { overlayRef, updatePopper, overlayProps } = useTooltipOverlay(props, config, mergedPrefixCls)
    const ellipsisProps = computed(() => props.ellipsis)

    expose({ updatePopper })

    return () => {
      const prefixCls = mergedPrefixCls.value
      let ellipsisContent = null

      if (ellipsisProps.value) {
        const wrap = isEllipsis(props.title!, ellipsisProps.value.width)
        if (!wrap) {
          return <span class={`${prefixCls}-ellipsis`}>{props.title}</span>
        }

        const style: CSSProperties = {
          display: 'inline-block',
          width: `${ellipsisProps.value.width}px`,
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          whiteSpace: 'nowrap',
        }

        ellipsisContent = (
          <span class={`${prefixCls}-ellipsis`} style={style}>
            {props.title}
          </span>
        )
      }

      return (
        <ɵOverlay
          ref={overlayRef}
          v-slots={{
            default: ellipsisProps.value ? ellipsisContent : slots.default,
            content: () => renderContent(props, slots, prefixCls),
          }}
          class={prefixCls}
          transitionName={`${common.prefixCls}-fade-fast`}
          {...overlayProps.value}
        />
      )
    }
  },
})

const renderContent = (props: TooltipProps, slots: Slots, prefixCls: string) => {
  if (!(slots.title || props.title)) {
    return null
  }
  return <div class={`${prefixCls}-wrapper`}>{slots.title ? slots.title() : props.title}</div>
}

// 判断是否换行
const isEllipsis = (title: string, width: number) => {
  const span = document.createElement('span')
  span.style.visibility = 'hidden'
  span.style.whiteSpace = 'nowrap'
  span.innerText = title
  document.body.appendChild(span)
  const length = span.offsetWidth
  document.body.removeChild(span)

  return length > width
}
