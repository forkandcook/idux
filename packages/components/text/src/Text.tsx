/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, h, normalizeClass, ref } from 'vue'

import { useClipboard } from '@idux/cdk/clipboard'
import { callEmit } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { IxIcon } from '@idux/components/icon'

import IxTypography from './Typography'
import { textProps } from './types'

export default defineComponent({
  name: 'IxText',
  props: textProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-text`)

    const container = ref<HTMLElement | null>(null)

    const copied = ref(false)
    const { copy } = useClipboard()
    function handleCopy(evt: Event) {
      evt.stopPropagation()
      if (copied.value) {
        return
      }

      copied.value = true
      const content = container.value?.innerText ?? ''
      copy(content).then(() => {
        setTimeout(() => {
          copied.value = false
        }, 3000)
      })
    }

    const handleClick = (evt: Event) => {
      if (props.onClick) {
        callEmit(props.onClick, evt)
      }
    }

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-${props.type}`]: !!props.type,
        [`${prefixCls}-disabled`]: props.disabled,
        [`${prefixCls}-ellipsis`]: props.ellipsis,
      })
    })

    const renderContent = () => {
      if (!slots.default) {
        return null
      }

      let content = slots.default()

      if (props.copyable) {
        content.push(
          <a class={`${mergedPrefixCls.value}-copy`} onClick={handleCopy}>
            {copied.value ? <IxIcon name="check" /> : <IxIcon name="copy" />}
          </a>,
        )
      }

      function wrapContent(tag: string, isDefine: boolean) {
        if (!isDefine) {
          return
        }

        content = [h(tag, content)]
      }

      wrapContent('code', props.code)
      wrapContent('kbd', props.keyboard)
      wrapContent('u', props.underline)
      wrapContent('del', props.delete)
      wrapContent('strong', props.strong)
      wrapContent('i', props.italic)
      wrapContent('mark', props.mark)

      return content
    }

    return () => {
      return props.ellipsis ? (
        <div class={classes.value} onClick={handleClick}>
          <IxTypography {...props} v-slots={slots}></IxTypography>
        </div>
      ) : (
        <span class={classes.value} onClick={handleClick} ref={container}>
          {renderContent()}
        </span>
      )
    }
  },
})
