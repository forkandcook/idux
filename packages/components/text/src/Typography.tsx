/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, onMounted, onUnmounted, ref, watch } from 'vue'

import { useClipboard } from '@idux/cdk/clipboard'
import { callEmit } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { IxIcon } from '@idux/components/icon'
import { IxTooltip } from '@idux/components/tooltip'

import { useEllipsisCSS } from './composables/useEllipsisCss'
import { useExpanded } from './composables/useExpanded'
import { useTextProps } from './composables/useTextProps'
import { textProps } from './types'
import { truncateText } from './utils'

export default defineComponent({
  props: {
    ...textProps,
  },
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-text`)

    const { width, rows, tooltip, onExpand } = useTextProps(props)
    const { expanded } = useExpanded(props)

    const copied = ref(false)
    const { copy } = useClipboard()

    function handleCopy(evt: Event) {
      evt.stopPropagation()
      if (copied.value) {
        return
      }
      copied.value = true
      const content = container.value?.innerText ?? ''
      copy(content).then(() => {
        setTimeout(() => {
          copied.value = false
        }, 3000)
      })
    }

    watch(expanded, () => {
      callEmit(onExpand)
      if (props.copyable && props.ellipsis) {
        truncateText(container.value!, domWidth.value * (rows ?? 1), expanded.value)
      }
    })

    const container = ref<HTMLElement | null>(null)
    const domWidth = ref(0)

    const handleResize = () => {
      domWidth.value = container.value ? container.value.getClientRects()[0].width : 0

      if (props.copyable && props.ellipsis) {
        truncateText(container.value!, domWidth.value * (rows ?? 1))
      }
    }

    onMounted(() => {
      handleResize()
      window.addEventListener('resize', handleResize)
    })

    onUnmounted(() => {
      window.removeEventListener('resize', handleResize)
    })

    const styles = useEllipsisCSS(width, rows)

    function render() {
      return (
        <a class={`${mergedPrefixCls.value}-copy`} onClick={handleCopy}>
          {copied.value ? <IxIcon name="check" /> : <IxIcon name="copy" />}
        </a>
      )
    }

    return () => {
      if (!slots.default) {
        return null
      }

      const content = slots.default()

      return tooltip && !expanded.value ? (
        <IxTooltip title={tooltip}>
          <p style={expanded.value ? undefined : props.copyable ? undefined : styles} ref={container}>
            <span>{content}</span>
            <span>{props.copyable && render()}</span>
          </p>
        </IxTooltip>
      ) : (
        <p style={expanded.value ? undefined : props.copyable ? undefined : styles} ref={container}>
          <span>{content}</span>
          <span>{props.copyable && render()}</span>
        </p>
      )
    }
  },
})
