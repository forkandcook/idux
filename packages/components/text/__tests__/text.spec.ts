import { MountingOptions, mount } from '@vue/test-utils'

import { renderWork } from '@tests'

import Text from '../src/Text'
import { TextProps } from '../src/types'

const TextMount = (options?: MountingOptions<Partial<TextProps>>) =>
  mount(Text, { ...(options as MountingOptions<TextProps>) })

renderWork<TextProps>(Text)

describe('Text', () => {
  test('type work', async () => {
    const wrapper = TextMount({ props: { type: 'secondary' } })
    expect(wrapper.classes()).toContain('ix-text-secondary')

    await wrapper.setProps({ type: 'success' })
    expect(wrapper.classes()).toContain('ix-text-success')

    await wrapper.setProps({ type: 'warning' })
    expect(wrapper.classes()).toContain('ix-text-warning')

    await wrapper.setProps({ type: 'danger' })
    expect(wrapper.classes()).toContain('ix-text-danger')

    await wrapper.setProps({ type: undefined })
    expect(wrapper.classes()).not.toContain('ix-text-secondary')
    expect(wrapper.classes()).not.toContain('ix-text-success')
    expect(wrapper.classes()).not.toContain('ix-text-warning')
    expect(wrapper.classes()).not.toContain('ix-text-danger')
  })

  test('disabled work', async () => {
    const wrapper = TextMount({ props: { disabled: true } })
    expect(wrapper.classes()).toContain('ix-text-disabled')

    await wrapper.setProps({ disabled: false })
    expect(wrapper.classes()).not.toContain('ix-text-disabled')
  })

  test('delete work', async () => {
    const wrapper = TextMount({ props: { delete: true } })
    expect(wrapper.find('del')).not.toBeNull()

    await wrapper.setProps({ delete: false })
    expect(wrapper.find('del').exists()).toBeFalsy
  })

  test('underline work', async () => {
    const wrapper = TextMount({ props: { underline: true } })
    expect(wrapper.find('u')).not.toBeNull()

    await wrapper.setProps({ underline: false })
    expect(wrapper.find('u').exists()).toBeFalsy
  })

  test('mark work', async () => {
    const wrapper = TextMount({ props: { mark: true } })
    expect(wrapper.find('mark')).not.toBeNull()

    await wrapper.setProps({ mark: false })
    expect(wrapper.find('mark').exists()).toBeFalsy
  })

  test('code work', async () => {
    const wrapper = TextMount({ props: { code: true } })
    expect(wrapper.find('code')).not.toBeNull()

    await wrapper.setProps({ code: false })
    expect(wrapper.find('code').exists()).toBeFalsy
  })

  test('keyboard work', async () => {
    const wrapper = TextMount({ props: { keyboard: true } })
    expect(wrapper.find('kbd')).not.toBeNull()

    await wrapper.setProps({ keyboard: false })
    expect(wrapper.find('kbd').exists()).toBeFalsy
  })

  test('strong work', async () => {
    const wrapper = TextMount({ props: { strong: true } })
    expect(wrapper.find('strong')).not.toBeNull()

    await wrapper.setProps({ strong: false })
    expect(wrapper.find('strong').exists()).toBeFalsy
  })

  test('italic work', async () => {
    const wrapper = TextMount({ props: { italic: true } })
    expect(wrapper.find('i')).not.toBeNull()

    await wrapper.setProps({ italic: false })
    expect(wrapper.find('i').exists()).toBeFalsy
  })

  test('ellipsis work', () => {
    const wrapper = TextMount({ props: { ellipsis: true } })
    expect(wrapper.classes()).toContain('ix-text-ellipsis')
  })

  test('copyable work', () => {
    const wrapper = TextMount({ props: { copyable: true } })
    expect(wrapper.find('.ix-text-copy')).not.toBeNull()
  })
})
