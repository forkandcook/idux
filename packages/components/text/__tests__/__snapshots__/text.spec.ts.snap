// Vitest Snapshot v1

exports[`render work 1`] = `"<span class=\\"ix-text\\"><!----></span>"`;
